Resolve input and output execution states explicitly

lptr opened the pull request:
The PR makes resolving a task's output and input states before execution explicit. This means that instead of snapshotting the inputs and outputs on demand when needed, these operations happen in specific places in the execution chain.

This allows two things:

1) moving more task executors to `:execution` in preparation for incremental/cacheable artifact transforms,
2) better capture the snapshotting times; we now know when snapshotting happens, and we can also measure the time it takes to snapshot inputs and outputs (before and after execution) separately. This helps resolve [this comment](https://github.com/gradle/gradle/blob/c58fe44922178adc6cfff2358c2e57ae14b520a3/subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveBuildCacheKeyExecuter.java#L83-L99) from @ldaley.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/IncrementalTaskAction.java:
```diff
@@ -41,9 +53,32 @@ public void releaseContext() {
+
+    private RebuildIncrementalTaskInputs createRebuildInputs(Task task) {
+        ImmutableCollection<CurrentFileCollectionFingerprint> currentInputs = context.getBeforeExecutionState().get().getInputFileProperties().values();
+        return instantiator.newInstance(RebuildIncrementalTaskInputs.class, task, currentInputs);
```
I wonder if we really need to use an `Instantiator` here?

lptr replied:
👍

lptr commented:
@ldaley please give some advice here.

Before this PR, we had quite a bit of mess around capturing the bookkeeping time for snapshotting stuff and calculating cache keys. We captured and measured a single build operation in `ResolveBuildCacheKeyExecuter` that included a few separate steps. We had to measure these steps together because they were done atomically. Here's your comment about it:

https://github.com/gradle/gradle/blob/ff0ed56cc64e224b7b028a07045d2a36158d957b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveBuildCacheKeyExecuter.java#L139-L155

In this PR each step now happens in its own separate class, and thus can be captured and measured separately, making that comment and the existing build operation obsolete.

- snapshot the task's _outputs before_ execution: `ResolveBeforeExecutionOutputsTaskExecuter`
- snapshot the task's _inputs before_ execution: `ResolveBeforeExecutionStateTaskExecuter`
- calculate the cache key from the _inputs:_ `ResolveBuildCacheKeyExecuter`
- snapshot the task's _output after_ execution: `SnapshotOutputStep`

(The `*Executer` types will move to the `:execution` project at some point, and become `*Step`s, but they will stay separate steps.)

How should we move forward with the current single build operation? The goal should be to measure these things separately indeed, but we need to keep backwards compatible, too. Please advise.

wolfs commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveBuildCacheKeyExecuter.java:
```diff
@@ -80,6 +135,7 @@ public TaskExecuterResult execute(TaskInternal task, TaskStateInternal state, Ta
     }
 
     private void resolve(final TaskInternal task, final TaskExecutionContext context) {
+        // TODO:lptr Review this comment
```
The inputs are now snapshotted in `ResolveBeforeExecutionStateTaskExecuter`. I guess there needs to be a change in the scan plugin, since we capture snapshotting time now as well. The time captured by this build operation is only the time it takes to calculate the build cache key from the fingerprints.

lptr replied:
Yes. We need some structural changes in tracking build operations to accommodate this. I'm talking to @ldaley about how to proceed on this.

lptr replied:
The structural changes are in.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveBuildCacheKeyExecuter.java:
```diff
@@ -114,15 +170,21 @@ public void run(BuildOperationContext buildOperationContext) {
+    private TaskOutputCachingBuildCacheKey doResolve(final TaskInternal task, TaskExecutionContext context) {
+        final TaskProperties taskProperties = context.getTaskProperties();
+        return context.getBeforeExecutionState()
+            .map(new java.util.function.Function<BeforeExecutionState, TaskOutputCachingBuildCacheKey>() {
```
Ah, right, I wanted to do that actually.

lptr replied:
OK, I checked it, and the problem is that you need to use Guava's `Function`s with their collections API. We could use the Java 8 stream API, and we should, but let's do this in a separate step at some point.

lptr replied:
BTW, one of the goals of this exercise is that we can move all this to `:execution`, which would allow us to use lambdas. That would at least hide the ugliness here, and it would also make it easier to switch to the stream API.

wolfs replied:
Since the usages of Guava's `Function` are now gone from this class, we can use an import already.

lptr replied:
Done.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/changedetection/TaskExecutionMode.java:
```diff
@@ -0,0 +1,62 @@
+ */
+public enum TaskExecutionMode {
+    INCREMENTAL(null, true, true),
+    WITHOUT_ACTIONS("Task has not declared any outputs nor actions.", false, false),
```
Good point.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/changedetection/TaskExecutionModeResolver.java:
```diff
@@ -18,6 +18,6 @@
-public interface TaskArtifactStateRepository {
-    TaskArtifactState getStateFor(TaskInternal task, TaskProperties taskProperties);
+public interface TaskExecutionModeResolver {
+    TaskExecutionMode getStateFor(TaskInternal task, TaskProperties taskProperties);
```
Good point.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java:
```diff
@@ -101,9 +112,10 @@ public TaskExecuterResult execute(TaskInternal task, TaskStateInternal state, fi
                 }
-                taskArtifactState.snapshotAfterTaskExecution(context);
             }
+            // TODO:lptr can we remove this?
```
Yeah, but if you follow the code, we are always passing in an empty `FileCollection` here. Still, my earlier attempts at removing it ended up with some mysterious failures.

wolfs replied:
Source files is a live file collection isn't it? So it's value can change. Mostly it will be a file tree for some directory, which will have files as soon as there are some files in the directory.

lptr replied:
I'm confused. All we do in `DefaultTaskInputsListener.onExecute()` is that we call `FileCollectionInternal.registerWatchPoints()`. And what that method does is it eagerly adds every `File` in the collection to a builder. There's nothing lazy/live about this: we take what files there are in the collection right now, which should be none at this point in the code anyway.

wolfs replied:
If the collection is a file tree, then `getFiles()` will return an empty list for an empty directory. Registering the watch point will end up [here](https://github.com/gradle/gradle/blob/b16479080a8864f91d15540d41bc09bdbe2c0af3/subprojects/files/src/main/java/org/gradle/api/internal/file/FileSystemSubset.java#L134) where we add the root directory of the file tree as a watch point.

wolfs replied:
Yeah, see my comment just above.

lptr replied:
Makes sense, thanks, removed the comment.

lptr commented on subprojects/core/src/main/java/org/gradle/execution/ProjectExecutionServices.java:
```diff
@@ -92,12 +94,15 @@ TaskActionListener createTaskActionListener(ListenerManager listenerManager) {
 
-    TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository,
+    TaskExecuter createTaskExecuter(TaskExecutionModeResolver repository,
                                     BuildCacheCommandFactory commandFactory,
```
I'll see to that.

One of the things I'd love to change for service registries is if we could somehow tell IntelliJ that the methods are used, so all the "unused" warnings reported would be actionable. By requiring an annotation on these types (`@Provides` comes to mind from Guice) could help achieve that. /cc @adammurdoch

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java:
```diff
@@ -227,12 +245,53 @@ public void outputsRemovedAfterFailureToLoadFromCache() {
+            AfterPreviousExecutionState afterPreviousExecutionState = context.getAfterPreviousExecution();
+            ImmutableSortedMap<String, CurrentFileCollectionFingerprint> outputFilesBeforeExecution = context.getOutputFilesBeforeExecution();
+
+            OverlappingOutputs overlappingOutputs = OverlappingOutputs.detect(
```
I'm planning to calculate overlapping outputs and put them in the context.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java:
```diff
@@ -227,12 +245,53 @@ public void outputsRemovedAfterFailureToLoadFromCache() {
+                    public void accept(BeforeExecutionState execution) {
+                        executionHistoryStore.store(
+                            task.getPath(),
+                            OriginMetadata.fromPreviousBuild(originMetadata.getBuildInvocationId(), originMetadata.getExecutionTime()),
```
We shouldn't, really. Good catch.

lptr replied:
OK, reverting this makes things fail in an arcane way, investigating...

lptr replied:
OK, so the problem is that without this re-wrapping we pass on an origin metadata object that is produced by the current build. With the wrapping we reset it to say it was produced by a previous build. This _should not_ matter, as all we do here is we serialize the object to be stored on disk: the next time we read it we will invariably set it to "produced by previous build".

However, caching is applied transparently via a decorator during runtime that stores the actual object, and returns it when requested in the next build. 🤦‍♂️ Thus we actually need to pass an object that we want to get back even though the store should "normalize" it via serialization, it doesn't happen when the cache gets a hit.

I'll revert to the original.

lptr replied:
OK, I ended up moving `isProducedByCurrentBuild()` to `SnapshotResult.isReused()` where it seems to make a lot more sense. That allowed me to remove this hack, too. All much cleaner now!

wolfs commented on subprojects/core/src/main/java/org/gradle/caching/internal/tasks/DefaultTaskCacheKeyCalculator.java:
```diff
@@ -0,0 +1,71 @@
+            builder.appendInputFilesProperty(entry.getKey(), entry.getValue());
+        }
+
+        for (TaskOutputFilePropertySpec propertySpec : taskProperties.getOutputFileProperties()) {
```
Why don't we use the names in `BeforeExecutionState.getOutputFileProperties` as output file propery names?

lptr replied:
I guess we could. Mind adding a card for this?

wolfs replied:
https://github.com/gradle/build-cache/issues/1388

wolfs commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveIncrementalChangesTaskExecuter.java:
```diff
@@ -0,0 +1,115 @@
+                @Nullable
+                @Override
+                public ExecutionStateChanges get() {
+                    final AfterPreviousExecutionState afterPreviousExecution = context.getAfterPreviousExecution();
```
Shouldn't `AfterPreviousExecution` also be an Optional? So we could use `filter` and `flatMap`?

lptr replied:
It probably should. But: baby steps! I think all these changes will be both more obvious and easier to make when have moved things over to `:execution`.

wolfs commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveTaskExecutionModeExecuter.java:
```diff
@@ -94,7 +94,7 @@ public String getDisplayName() {
         } finally {
             outputs.setPreviousOutputFiles(null);
-            context.setTaskArtifactState(null);
+            context.setTaskExecutionMode(null);
```
Why do we sometimes clear things from the context (like here) and don't in other places (like `ResolveBeforeExecutionStateTaskExecuter`)?

wolfs replied:
Maybe we do the cleaning so we ensure no one up the chain accesses task artifact state?

Now that the arguments of the setters are marked as non null, we also get warnings here about passing in `null`. Shall we remove the clearing?

lptr replied:
This is not a regression in this PR. I also intend to completely revamp this when moving to `:execution` in a way that makes access to unavailable data impossible. I don't want to touch this stuff until then.

wolfs commented on subprojects/core/src/main/java/org/gradle/execution/ProjectExecutionServices.java:
```diff
@@ -115,24 +120,31 @@ TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository,
         executer = new ValidatingTaskExecuter(executer);
         executer = new SkipEmptySourceFilesTaskExecuter(inputsListener, executionHistoryStore, cleanupRegistry, outputChangeListener, executer, buildInvocationScopeId);
-        executer = new ResolvePreviousStateExecuter(executionHistoryStore, executer);
+        executer = new ResolveBeforeExecutionOutputsTaskExecuter(fingerprinterRegistry, executer);
```
I guess we only need the output snapshots before the `SkipEmptySourceFilesTaskExecuter` if we have a task history. If we don't, there is no need to detect overlapping outputs for cleanup there.

lptr replied:
What are you suggesting to change here?

wolfs replied:
That we could only resolve the before execution outputs when required in `SkipEmptySourceFilesTaskExecuter`. That would require some laziness for resolving the execution outputs again - so I guess we shouldn't do it.

The change to not resolve the outputs when there are no source files and no previous task history could speed up that step.

lptr replied:
How would that be useful? I mean, the only reason why we _wouldn't_ observe the outputs in `SkipEmptySourceFilesTaskExecuter` is if we continue executing the task further down the line. On that path the only case where we can finish without needing the output snapshots is if `ValidatingTaskExecuter` fails. That's a case I don't think is worth optimizing for.

wolfs replied:
We also don't need to observe the outputs if there is no task history and no sources. So the optimization would be for the no-source case.

lptr replied:
What we could do is split the empty case check into two, and do this special case before. But I don't think it's worth it at this point. If you do, please create an issue so we can discuss in detail.

wolfs commented on subprojects/core/src/main/java/org/gradle/execution/ProjectExecutionServices.java:
```diff
@@ -115,24 +120,31 @@ TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository,
+            executer = new ResolveBuildCacheKeyExecuter(buildOperationExecutor, cacheKeyCalculator, buildCacheController.isEmitDebugLogging(), executer);
         }
+        executer = new ResolveBeforeExecutionStateTaskExecuter(classLoaderHierarchyHasher, valueSnapshotter, fingerprinterRegistry, executer);
         executer = new ValidatingTaskExecuter(executer);
```
`ValidatingTaskExecuter` could re-use the input snapshots from `ResolveBeforeExecutionStateTaskExecuter` instead of checking manually for existence of files and types, right?

lptr replied:
Sounds right, would you mind adding a card to do this?

wolfs replied:
https://github.com/gradle/build-cache/issues/1389

lptr replied:
Thanks!

lptr commented:
I've added a test for `ResolveBeforeExecutionOutputsTaskExecuter`, but it's pretty trivial. What cases do you think we should test for `ResolveBeforeExecutionStateTaskExecuter`?

lptr commented:
@ldaley I've added the non-wrap-around `BuildOperation` feature here: https://github.com/gradle/gradle/pull/8038/commits/705a42270926100a03fc3710422100f6029e2e57

It doesn't yet check if all child operations are finished, but I think the approach is already worth reviewing. Could you take a look?

ldaley commented:
Regarding: https://github.com/gradle/gradle/pull/8038/commits/ea106fcba012a7fac2b6c7429f6782800892dcbf

1. The duplication in the executor should be dealt with before merge or immediately after, otherwise it _will_ diverge.
2. I didn't find a test for the handling of bugged code not finishing the child op before the parent finishes
3. We should have an understanding of what will happen if we ship with a bug that means that child tries to outlive the parent - does the build fail reasonably? Or does this cause a cascade of failures that is entirely unintelligible? The latter might still be OK, but I think we should know this.

wolfs commented:
Regarding tests for `ResolveBeforeExecutionStateTaskExecuter`: I would expect tests that the fingerprints are correctly created from the taskProperties and the task class/actions. If we have other tests which already check this, then we don't need to add more tests.

The handling of `isTaskHistoryMaintained` should probably also be tested.

lptr commented on subprojects/execution/src/test/groovy/org/gradle/internal/execution/impl/steps/CacheStepTest.groovy:
```diff
@@ -72,8 +72,9 @@ class CacheStepTest extends Specification {
     def "executes work and stores in cache on cache miss"() {
```
Good catch!

lptr commented on subprojects/execution/src/test/groovy/org/gradle/internal/execution/impl/steps/IncrementalExecutionTest.groovy:
```diff
@@ -777,7 +777,7 @@ class IncrementalExecutionTest extends Specification {
                 @Override
                 Optional<ExecutionStateChanges> getChangesSincePreviousExecution() {
-                    changes = Optional.ofNullable(executionHistoryStore.load(getIdentity())).map { previous ->
+                    changes = executionHistoryStore.load(getIdentity()).map { previous ->
```
👍

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/DefaultTaskExecutionContext.java:
```diff
@@ -62,13 +75,62 @@ public void setAfterPreviousExecution(@Nullable AfterPreviousExecutionState afte
+    }
+
+    @Override
+    public Optional<OverlappingOutputs> getOverlappingOutputs() {
```
We already have some error checking: you'd be getting `NPE`s if you asked for stuff too early. But this is exactly the situation I'd like to remedy by moving to an immutable approach soon.

lptr commented:
@wolfs I've addressed your comments. For testing `ResolveBeforeExecutionStateTaskExecuter` I've added https://github.com/gradle/build-cache/issues/1391.
